### PR TITLE
Add WebSocket client support

### PR DESCRIPTION
**Pull request by erikcorry**

### CODE REVIEW

File: src/web_socket.toit [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+TWO_BYTE_SIZE_       ::= 126
+MAX_ONE_BYTE_SIZE_   ::= 125
+
+/**
+A WebSocket connection.
+A bidirectional socket connection capable of sending binary or text messages
+  according to RFC 6455.
+Obtained from the $Client.web_socket method.
+Currently the server side setup is not implemented.
+Currently does not implement ping and pong packets.
```
Comment (floitsch) [inline src/web_socket.toit:27]: I feel like this doesn't really belong into the toitdoc, but it probably doesn't hurt. (We might forget to remove the lines when we add the functions, though).

File: src/web_socket.toit [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+    return socket_.read
+
+  unread_ byte_array/ByteArray -> none:
+    assert: pending_ == null
+    pending_ = byte_array
+
+  /**
+  Returns a reader for the next message sent to us on the WebSocket.
+  Returns null if the connection is closed.
+  Should not be called until the previous reader has been fully read.
```
Comment (floitsch) [inline src/web_socket.toit:51]: The alternative to guarantee this, is to use a block:
`receive [block] -> none`

Also, I wouldn't be surprised if you often just want to get one frame.
Maybe adding a `receive -> any` that just does a read-all and gives the string or byte-array to you?

Obvious question is then, whether the function should be called `read` then. Maybe `read_frame` ?

File: src/web_socket.toit [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+  */
+  receive -> WebSocketReader?:
+    if current_reader_ != null: throw "PREVIOUS_READER_NOT_FINISHED"
+    fragment_reader := receive_
+    if fragment_reader == null: return null
+    if fragment_reader.is_ping or fragment_reader.is_pong:
+      throw "UNIMPLEMENTED_PING"
```
Comment (floitsch) [inline src/web_socket.toit:58]: Not saying this is important, but I would expect that implementing 'pong' should be trivial.

File: src/web_socket.toit [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+    into multiple fragments, which would ruin this.  In practice this seems
+    rare, and with TLS it is almost impossible.
+  The fragment size is at the WebSockets level and has no connection with
+    the size of IP packets or TLS buffers.
+  The message is sent as a text message if the first data written to the
+    writer is a string, otherwise as a binary message.
+  Returns a writer, which must be completed (all data sent, and closed) before
+    this method can be called again.
+  */
+  send --size/int?=null -> WebSocketWriter:
```
Comment (floitsch) [inline src/web_socket.toit:126]: Bad function name.

Maybe `start_sending` or `initiate_sending`, or `start_write`.

File: src/web_socket.toit [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+    written := 0
+    while written < data.size:
+      written += socket_.write data[written..]
+
+  reader_close_ -> none:
+    current_reader_ = null
+
+  /**
+  Abruptly closes the WebSocket.
+  Does not send a close packet first.
```
Comment (floitsch) [inline src/web_socket.toit:145]: How would you send a 'close' packet?

If there is a nice way to close (with a close packet), and a not-so-nice way, then I tend to have:
```
close --force/bool=false
```

Obviously depends on how it's done underneath.

File: src/client.toit [hunk 6]
```diff
@@ -182,6 +185,53 @@ class Client:
 
     throw "Too many redirects"
 
+  web_socket host/string --port/int?=null path/string --headers=Headers --follow_redirects/bool=true -> WebSocket:
```
Comment (floitsch) [inline src/client.toit:188]: Needs comments.

File: src/web_socket.toit [hunk 7]
```diff
@@ -0,0 +1,405 @@
...
+  */
+  close:
+    socket_.close
+    current_writer_ = null
+    current_reader_ = null
+
+  static add_client_upgrade_headers_ headers/Headers -> string:
+    nonce := base64.encode (ByteArray 16: random 0x100)
```
Comment (floitsch) [inline src/web_socket.toit:198]: Add a note that the nonce isn't super important and doesn't need to by cryptographically safe.

File: src/web_socket.toit [hunk 8]
```diff
@@ -0,0 +1,405 @@
...
+        or (Headers.ascii_normalize_ connection_header) != "Upgrade"
+        or (response.headers.single "Sec-WebSocket-Accept") != (WebSocket.response_ nonce):
+      throw "MISSING_HEADER_IN_RESPONSE"
+    if response.headers.single "Sec-WebSocket-Extensions"
+        or response.headers.single "Sec-WebSocket-Protocol":
+      throw "UNKNOWN_HEADER_IN_RESPONSE"
+
+  /**
+  Checks whether the request is a WebSocket upgrade request.
+  If it is a valid upgrade request, adds the required headers to the response_writer
```
Comment (floitsch) [inline src/web_socket.toit:223]: Document what the response is if it is an upgrade request: the nonce that was given by the client.

File: src/web_socket.toit [hunk 9]
```diff
@@ -0,0 +1,405 @@
...
+  Otherwise responds with an error code and returns null.
+  */
+  static check_server_upgrade_request_ request/Request response_writer/ResponseWriter_ -> string?:
+    connection_header := request.headers.single "Connection"
+    upgrade_header := request.headers.single "Upgrade"
+    version_header := request.headers.single "Sec-WebSocket-Version"
+    nonce := request.headers.single "Sec-WebSocket-Key"
+    message := null
+    if nonce == null:                message="No nonce"
+    else if nonce.size != 24:        message="Bad nonce size"
```
Comment (floitsch) [inline src/web_socket.toit:234]: Either align the `message =` with the other ones or move them to the left.
Also:
```suggestion
    if nonce == null:                message = "No nonce"
    else if nonce.size != 24:        message = "Bad nonce size"
```
Here and for the other 3.

File: src/web_socket.toit [hunk 10]
```diff
@@ -0,0 +1,405 @@
...
+      response_writer.headers.add "Connection" "upgrade"
+      response_writer.headers.add "Upgrade" "websocket"
+      return nonce
+    response_writer.write_headers STATUS_BAD_REQUEST --message=message
+    return null
+
+  static response_ nonce/string -> string:
+    expected_response := base64.encode
+        sha1 nonce + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
+    return expected_response
```
Comment (floitsch) [inline src/web_socket.toit:249]: This is more confusing on its own than when it is inlined.
Alternatively (to inlining), make this function responsible for upgrading:
```suggestion
  static write_upgrade_response_ response_writer/ResponseWriter_ nonce/string -> string:  
    expected_nonce_response := base64.encode
        sha1 nonce + "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
     response_writer.headers.add "Sec-WebSocket-Accept" expected_nonce_response
     response_writer.headers.add "Connection" "upgrade"
     response_writer.headers.add "Upgrade" "websocket"
     return nonce
```

And then in `check_server_upgrade_`:
```
  return write_upgrade_response_ response_writer nonce
```

File: src/web_socket.toit [hunk 11]
```diff
@@ -0,0 +1,405 @@
...
+  fragment_sent_ /bool := false
+
+  constructor.private_ .owner_ .size_:
+
+  write data from=0 to=data.size -> int:
+    if owner_ == null: throw "ALREADY_CLOSED"
+    total_size := to - from
+    while true:
+      while from < to and remaining_in_fragment_ != 0:
+        size := min to - from remaining_in_fragment_
```
Comment (floitsch) [inline src/web_socket.toit:267]:
```suggestion
        size := min (to - from) remaining_in_fragment_
```

File: src/web_socket.toit [hunk 12]
```diff
@@ -0,0 +1,405 @@
...
+          header[1] = TWO_BYTE_SIZE_
+          BIG_ENDIAN.put_uint16 header 2 size_
+        else:
+          header = ByteArray 2
+          header[1] = size_
+        header[0] = opcode | FIN_FLAG_
+      else:
+        // We don't know the size.  Write multiple fragments of up to
+        // 126 bytes.
+        remaining_in_fragment_ = min MAX_ONE_BYTE_SIZE_ data.size - from
```
Comment (floitsch) [inline src/web_socket.toit:302]:
```suggestion
        remaining_in_fragment_ = min MAX_ONE_BYTE_SIZE_ (data.size - from)
```

File: src/web_socket.toit [hunk 13]
```diff
@@ -0,0 +1,405 @@
...
+  remaining_in_fragment_ /int := 0
+  fragment_sent_ /bool := false
+
+  constructor.private_ .owner_ .size_:
+
+  write data from=0 to=data.size -> int:
+    if owner_ == null: throw "ALREADY_CLOSED"
+    total_size := to - from
+    while true:
+      while from < to and remaining_in_fragment_ != 0:
```
Comment (floitsch) [inline src/web_socket.toit:266]: Reorder this function.
The header is sent first, so it should be on the top of the while loop.
It's really confusing to first see the code for the body, and then only the header.

File: src/web_socket.toit [hunk 14]
```diff
@@ -0,0 +1,405 @@
...
+class WebSocketReader implements reader.Reader:
+  owner_ /WebSocket? := ?
+  is_text /bool
+  fragment_reader_ /FragmentReader_ := ?
+
+  /**
+  If the size of the incoming message is known, then it is returned, otherwise
+    null is returned.
+  Since a normal size method cannot return null, the WebSocketReader does not
+    implement $reader.SizedReader.
```
Comment (floitsch) [inline src/web_socket.toit:330]: You could create a subclass that implements SizedReader (and just overwrites the `size` field to return `-> int`).
An `is SizedReader` would then work.

Really not sure it's worth it.

File: src/web_socket.toit [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+  is_exhausted -> bool:    return received_ == size_
+
+  is_ok_ -> bool:
+    if control_bits_ & 0x70 != 0: return false
+    opcode := control_bits_ & 0xf
+    return opcode < 3 or 8 <= opcode <= 10
+
+  read -> ByteArray?:
+    if received_ == size_:
+      return null
```
Comment (floitsch) [inline src/web_socket.toit:387]:
```suggestion
    if received_ == size_: return null
```

File: src/web_socket.toit [hunk 16]
```diff
@@ -0,0 +1,405 @@
...
+      // TODO: Use blit for this.
+      masking_bytes.size.do:
+        next_byte_array[it] ^= masking_bytes[received_++ & 3]
+    else:
+      received_ += next_byte_array.size
+    return next_byte_array
+
+  size -> int?:
+    if control_bits_ & FIN_FLAG_ == 0:
+      return null
```
Comment (floitsch) [inline src/web_socket.toit:404]:
```suggestion
    if control_bits_ & FIN_FLAG_ == 0: return null
```

File: tests/websocket_standalone_test.toit [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+// Use of this source code is governed by a Zero-Clause BSD license that can
+// be found in the EXAMPLES_LICENSE file.
+
+import expect show *
+import http
+import net
+
+// Sets up a web server that can switch to websocket mode on the "/" path.
+// The server just sends back everything it gets.
+// Sets up a client that sends files and expects to recieve them back.
```
Comment (floitsch) [inline tests/websocket_standalone_test.toit:11]:
```suggestion
// Sets up a client that sends files and expects to receive them back.
```

File: src/web_socket.toit (hunk 10 above)
Comment (erikcorry) [reply src/web_socket.toit:249]: It's used two places, so can't inline, and can't really put more functionality in it.

File: src/web_socket.toit (hunk 8 above)
Comment (erikcorry) [reply src/web_socket.toit:223]: There's a bunch of headers that need to be right - I think the RFC documents this just fine.

File: src/web_socket.toit (hunk 12 above)
Comment (erikcorry) [reply src/web_socket.toit:302]: Should be `(to - from)`!

**Pull request merged**